### ISSUE TITLE
Null hypothesis significance testing - Z-Test / Gauss-Test

### ISSUE THREAD

**Issue by longmakesstuff** [scipy/scipy#13662]
**Is your feature request related to a problem? Please describe.**
I can not find a function to do a Z-Test in https://docs.scipy.org/doc/scipy/reference/stats.html. 

For reference of Z-Test: https://en.wikipedia.org/wiki/Z-test

**Describe the solution you'd like**
I would want to make a feature request (which I may want to contribute myself). I just want to be sure, that the feature is not yet planned.

**Comment by longmakesstuff** [comment]
@mdhaber The current textbook I am studying is unfortunately in german and I cannot find any (a bit more rigorous) reference, but this site provides a very good example https://www.sophia.org/tutorials/how-to-find-a-p-value-from-a-z-test-statistic-2. 

I can also try to describe the general approach to the test. Basically the Z-Test is quite similar to the T-Test, the only difference is, that in Z-Test we know the theoretical variance and hence use the normal distribution to compute the P values for following questions:

- Do the mean of the sample differs from the expected mean?
- Do the means of two independent samples differ? 
- Do the means of the same sample differ before and after?

I took a look at the T-Test implementations of Scipy and tried to implemented Z-Test myself, but I am not really familiar with Scipy's code standard and conventions. However I think, the implementation would not take that long for someone who have worked with Scipy's core modules.

**Comment by mdhaber** [comment]
It would not be hard to implement. Most of the time would go into documentation and testing. If you're not comfortable writing the code, you could help by designing the interface, drafting documentation, and writing tests that should pass once the function is implemented. 

Is this function implemented in other languages (e.g. R, Matlab, SASS)? Does it ever go by another name? Can you give a particular example of a situations it is used in? I would suggest emailing the SciPy developer mailing llist with this sort of information to see if others are interested in such a function being in SciPy.

**Comment by charlotte12l** [comment]
Hi, if no one is currently working on it, I would like to help implement the Z-test! I'm familiar with hypothesis significance testing :)

**Comment by mdhaber** [comment]
Hi @charlotte12l , I don't think there is anyone working on it. All new functions need an email to the [mailing list](https://mail.python.org/mailman/listinfo/scipy-dev) before they can be merged, so you might want to email the mailing list first to see if there is enough interest in such a function before starting. You could also email the list after submitting a PR, but then there's the chance that there's not interest.

Ideally, the email would include information like examples of when the test is used and evidence that the test is popular (e.g. highly cited paper). 

When you get started, I'd recommend taking a look at a recent PR, gh-12873 for an example of a new hypothesis test with a good example, good tests, and a really readable implementation. 

Also, since the Z-test uses the normal distribution, you can use [`_normtest_finish`](https://github.com/scipy/scipy/blob/3f017716532001349ccddb05498408d711bacdde/scipy/stats/stats.py#L1491) to get the p-value from the Z statistic. gh-13008 shows an example use, and it also has good testing and documentation of the `alternative` parameter.

**Comment by charlotte12l** [comment]
Thanks for your suggestions! That's very helpful! I'll write a proposal first this weekend and email it to the mailing list. :)

**Comment by ilayn** [comment]
Hi @charlotte12l, in #13938 another attempt has been started hence would you like to join forces or review or any other way be involved if you can still spare some time ?

**Comment by charlotte12l** [comment]
Sorry for the delayed reply. Sure, I'll join the reviews!

**Comment by angadgill** [comment]
Hi! This issue’s been open a while and looks like there was some progress on a PR but it was abandoned.

A Z Test is a very common hypothesis test and I’m surprised to see that it’s not part of Scipy. I can work on completing this!

If someone can let know if there’s any interesting in completing this, I would appreciate it. Thank you!

**Comment by mdhaber** [comment]
As mentioned [above](https://github.com/scipy/scipy/issues/13662#issuecomment-805539795), all new features need to be approved on the discussion forum.
There was a [post on the mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/APKLXGBNTRXOXX3GIACLCK773OWWOPKW/#APKLXGBNTRXOXX3GIACLCK773OWWOPKW) in 2021, but there wasn't much interest, so it wasn't added.
We now use a different [discussion forum](https://discuss.scientific-python.org/c/contributor/scipy/). You could try posting again, citing the first conversation and finding out whether there are other users who are interested. Please also make sure that there is a maintainer who is willing to review the PR before beginning work. If there is nobody who indicates that they are willing to review, it is likely that the PR will sit unreviewed for a long time.

I would normally suggest that we could allow the `ttest_1samp`, `ttest_rel`, and `ttest_ind` functions to accept $\sigma$ arguments and, if so, perform a z-test instead of a t-test, but I think those functions are already too bloated. I'd rather start fresh and keep it simple, following modern practices.

---

Based on recent activity in `scipy.stats`, I would be the most likely person to review. If you'd like me to review it, I'd ask you to write out some things here in the issue before opening a PR.

First, please find a text reference that is very explicit about the steps for both one- and two-sample versions of the test. It's not very challenging to invent the z-test with a little knowledge of statistics, but it's easier to review if there is just a recipe that's already laid out. 


Second, please find a reference implementation against which we can check the _output_ of the function. *Note - we cannot refer to any license-incompatible (e.g. GPL) code at all. So it's OK to pick an R implementation as a reference implementation; just don't look at the source.*

Next, suggest signatures of the test functions.  I'm assuming they would be named according to the convention established by  the t-test functions: `ztest_1samp`, `ztest_rel`, and `ztest_ind`. Let's not get too creative here since those are such well established functions. I think they would accept the sample(s) as the first positional argument(s) with `axis`, `nan_policy`, `keepdims`, and `alternative` as keyword arguments. You would need to suggest how the known standard deviations (or variances) would be accepted. I'd suggest that we not consider the possibility of weighted variants, trimming, or `method` options at this time. Also mention what attributes the result object (like `TtestResult`) would have in addition to `statistic` and `pvalue` attributes, If `confidence_interval` is something you'd like to see, let's save it for a follow-up PR rather than trying to cram it into the first PR. There is a nonlinear relationship between the size of a PR and the amount of time it typically takes to review.

Let's see how that goes and we can take it from there. Again, I'd recommend waiting before you start on implementation. There are some important SciPy-specific things you'd need to be aware of for this to go smoothly, so let's take it step by step. The implementation is probably going to be the least time-consuming part, maybe 5-10% of the work.

**Comment by angadgill** [comment]
@mdhaber -- Thank you SO much for your detailed response and your support. I appreciate it. 

I have created a [post over at the forum](https://discuss.scientific-python.org/t/proposal-add-z-test-to-scipy/2151) and will let you know how that goes. 

> Second, please find a reference implementation against which we can check the output of the function

Can [statsmodel](https://www.statsmodels.org/stable/generated/statsmodels.stats.weightstats.ztest.html) function as reference here?

> Next, suggest signatures of the test functions.

Naming them like the `ttest` functions makes a lot of sense. Perhaps an argument like `popvar` can  be used for cases when the population variance is known. I'll work on this and post back.

**Comment by mdhaber** [comment]
> Can statsmodel function as reference here?

If a does the same thing, sure. But that function doesn't accept the population variance(s). (Maybe it just assumes the sample variance is the population variance and uses the normal instead of t-distribution?)

**Comment by josef-pkt** [comment]
The main motivation for z-test (in statsmodels) is asymptotic where we need only assumption for the law of large numbers and central limit theorem to hold.
For example it works for binomial, proportion z-test, ...

The z-test has the exact distribution if the random variable is normally distributed and the variance is known. But those are strong assumptions that I don't think apply very often. (*)

I don't know of a reference, it's very standard and should be in basic statistics text books, but I have not looked at those in a long time.

I agree that an option for the user to provide the population variance would be useful.

(*) although our tests for poisson rates and binomial proportions with methods that have good small sample properties also assume that the distribution poisson or binomial is correctly specified, or at least that the variance is the same as in those distributions. (no excess dispersion as in negative-binomial or beta-binomial)

**edit**
Two asides:

- example for assumptions: The proportion z-test has valid inference under the assumption that there is no excess dispersion. If we use the standard z-test with estimated variance, then it does not require the no excess dispersion assumption.

- t distribution versus normal distribution for inference: The central limit theorem only provides a justification for the normal distribution of the test statistic. However, in small samples the t-distribution is often or most (?) of the time a better approximation to the small sample distribution. So t-tests are still preferable to the z-test even if we use asymptotic methods. https://github.com/statsmodels/statsmodels/issues/6708 (But if we have a really large sample z or t does not matter, and in really small sample both might still over-reject.)

Those were the motivation in statsmodels to add z-test next to t-test, and, reversed, add use_t = True option to models (like GLM and MLE models) that usually only use the asymptotic normal distribution.

and another aside (just for the fun of it):
A compromise between estimating and fixing the variance would be to shrink the estimated variance towards the prior variance or towards variance from other samples. It's not yet available in statsmodels https://github.com/statsmodels/statsmodels/issues/8397

**Comment by angadgill** [comment]
Hey @mdhaber! It's been about a week since I posted on the [discussion forum](https://discuss.scientific-python.org/t/proposal-add-z-test-to-scipy/2151) and there's been no activity on it. Just checking-in to see where that leaves this project. I'm still motivated to work on this! Are you still open to reviewing the PR? I'm unclear on how reviews work, would we need someone from the forum to also chime in?

**Comment by mdhaber** [comment]
In the absence of objections, it's possible for things to be added with limited support. I think it fits well in principle with the rest of `stats`. In some sense, it's more "fundamental" than a t-test (albeit probably a lot less used/useful). For me reviewing, it depends a bit on the prior art - maybe start with the items I suggested in https://github.com/scipy/scipy/issues/13662#issuecomment-3419061746 and we'll go from there.

**Comment by angadgill** [comment]
Hey @mdhaber! I've been working on these and have updates!

> First, please find a text reference that is very explicit about the steps for both one- and two-sample versions of the test.

These took a while and I found a few lectures / notes from universities:

1. [Lecture 16: Hypothesis Testing by Ye Tian, Columbia University](https://www.columbia.edu/~yt2661/S1201/slides/lecture-16.pdf?utm_source=chatgpt.com). This covers step-by-step instructions for all combinations of Z-Test we care about: 1 / 2 sample, 1 / 2 sided, known / unknown variance, and proportions. This also has numeric examples we can use for unit tests. I quite like this!
2. [Lecture 5: Hypothesis testing: One- and two-sample cases by Ernesto F. L. Amaral, Texas A&M University](https://www.ernestoamaral.com/docs/soci600-22fall/Lecture05.pdf?utm_source=chatgpt.com). Very similar to the previous source, has step-by-step instructions and numeric examples. Though the instructions aren't as explicit as 1. 
3. [The Z-Test, Washington University](https://courses.washington.edu/psy315/tutorials/z_test_tutorial.pdf) Only covers 1-sample Z-Test of means (no proportions or 2-sample), so not as complete as the other sources but provides more numeric examples if we need the of unit tests.

I didn't find references in textbooks that was as clearly laid out. Let me know if these look okay or if I should look for references of a different flavor!

> Second, please find a reference implementation against which we can check the output of the function.

These two `stastmodel` functions should cover what we need, except z-test of means with known variance:
- `statsmodels.stats.weightstats.ztest` ([link](https://www.statsmodels.org/stable/generated/statsmodels.stats.weightstats.ztest.html#statsmodels-stats-weightstats-ztest)): 1 and 2 sample z-test of means with unknown variance.
- `statsmodels.stats.proportion.proportions_ztest`([link](https://www.statsmodels.org/stable/generated/statsmodels.stats.proportion.proportions_ztest.html#statsmodels.stats.proportion.proportions_ztest)): 1 and 2 sample z-test of proportions with known and unknown variance.

This R function can fill in the gap for _known variance_ case:
- `z.test`([link](https://www.rdocumentation.org/packages/BSDA/versions/1.2.2/topics/z.test)): 1 and 2 sample z-test of means with known variance.

> Next, suggest signatures of the test functions. 

I would love your input here! I'm thinking that we add 4 functions:
- `ztest_1samp(a, popmean, popvar=None, axis=0, nan_policy='propagate', alternative='two-sided', keepdims=False)`
    - This will cover 1 sample z-test with 1 or 2 sided + known or unknown variance. `popvar` will be None (default) when population variance is unknown and `float` when known. This one is the most clear to me.
- `ztest_ind(a, b, popvar=None, axis=0, equal_var=True, nan_policy='propagate', alternative='two-sided', keepdims=False)`
    - This one is unclear to me, I'll do more reading on it. I'm not sure if "known variance" case for a 2-sample Z-Test exists / is needed. I'm also unsure if `equal_var` is needed here.
- `ztest_1samp_prop`. This is interesting, there are two ways we could input the proportions:
    - `ztest_1samp_prop(k, n, ...)`: separate ints for numbers and denominator, similar to `binomtest`
    - `ztest_1samp_prop(a, ...)`: single float for the proportion. This is similar to the other `ztest` and `ttest` functions so I'm leaning towards this option. The full signature would look like identical to `ztest_1samp`
- `ztest_ind_prop`. Two options here too and I'm leaning towards the second:
    - `ztest_1samp_prop(k, n, ...)`: separate int _lists_ for numbers and denominator, one element for each "side". Similar to `statsmodels.stats.proportion.proportions_ztest`
    - `ztest_ind_prop(a, b, ...)`: two floats, one each for the two sides. The full signature would look identical to `ztest_ind`

For all fo these, the `result` would be a new object almost identical to `TtestResult`:
- `ZTestResult`. Parameters: `statistic` and `pvalue`

How does all this look so far? Looking forward to your feedback!

**Comment by mdhaber** [comment]
Thanks for looking into this.

I don't understand the motivation for offering the means tests when the population variance is unknown. That made sense when people were using tables and performing tests manually. We have computers now, and the t-test is just better in this situation, right?

What are the proportion tests - also using the normal distribution as an approximation, but now as an approximation to the binomial distribution?

I thought the need was for a z-test when population variances *are* known. I assumed users had examples in which that condition is met (ChatGPT does), and if so, it would be reasonable to fill that gap. Do you have use cases where the variance/variances is/are actually known (or maybe estimated from *much* more data than is being used to compare the means in question)?

The `z.test` reference implementation looks perfect.

I don't actually see the theory for the two-sample case (with two known variances) in the references. (two-sided $\neq$ two-sample). The second might (page 56), but I'm not certain those standard deviations are known. They are written with $s$, not $\sigma$, for instance, and I would expect $n-1$ in an unknown variance case.  I think "pooled" typically refers to an estimate of a common variance, but we could have two different variances. However, I think the formula would be correct and what we want with known population variances $\sigma_i$ replacing $s_i$ and with $n_i$ instead of the $n_i-1$. That would be the standard error of the difference between two means. (And that is the standard deviation of the normal distribution that describes the difference between two sample means. We can perform the test by comparing the observed difference between sample means against that distribution, or equivalently, by normalizing the observed difference between sample means by that standard deviation, considering that to be the test statistic, and comparing against the standard normal.)

Anyway, even if we know the math, would be nice to have a reference we could include in the documentation - is it at one of those links?

**Comment by josef-pkt** [comment]
wikipedia article references 3 text books
https://en.wikipedia.org/wiki/Z-test

The last section in the wikipedia page is on two sample proportion z-test

The one sample proportion z-test can be derived through normal approximation of the binomial distribution.
But I didn't know about this until we got a stackoverflow question (criticizing our default variance)
(discussion to this answer https://stats.stackexchange.com/a/447599/14187 with reference to Zar text book)
For me, the usual explanation is based on large sample asymptotics.

In binomial proportion test, the variance depends on the mean. In the one sample test, the variance of the test statistic is fixed under the null hypothesis.
The same holds for the test for one sample poisson rates test when using a score test.

AFAIR, almost no article or book uses t distribution instead of normal distribution when testing poisson rates and binomial proportions.

**Comment by mdhaber** [comment]
@josef-pkt so what is your opinion on the utility of the proposed to tests compared to our existing offerings? If  you're advocating for the two sample proportion test, when is that preferred to Poisson means?

**Comment by josef-pkt** [comment]
I'm just arguing in favor of adding the proposed z-test functions.
Even though in many cases the t-test should be preferred, z-tests are basic tests that should be in a stats package.

My references to other distributions was mainly intended to show that using z-tests is very common, even though it's not often used in the continuous mean testing.
(My functions for proportions, rates and similar delegate to a generic z-test.)

**Comment by mdhaber** [comment]
Ok, but my motivation to work on this would need to be because it is useful, not just for completeness.

**Comment by josef-pkt** [comment]
google search AI for "popularity of z-test" says

"
The z-test is a frequently used and popular statistical tool in both research and industry, particularly in specific scenarios such as A/B testing and quality control.
"
another query "examples for z-test usage" gives more info

While looking for examples, I got distracted by a model comparison (LR) test which statsmodels does not have yet. That test can be implemented by just calling a standard paired z test (with estimated variance) with the appropriate test variables.
https://github.com/statsmodels/statsmodels/issues/3897#issuecomment-1324288326

So, it looks like that there are many applications for the z-test even if the standard hypothesis tests for continuous mean are dominated by the t-test.

LLMs are great (even if the details are sometimes incorrect)

copilot is more critical:
"
The z-test remains widely used in statistics, especially for large-sample hypothesis testing, but its popularity has declined relative to the t-test due to stricter assumptions.
"

:)

**Comment by angadgill** [comment]
My two cents from industry experience, specially in AB Testing: Z-Test is useful in all its forms but to varying degree. I would order them roughly as follows (most to least useful):

1. **2-sample Z-Test of Proportions, with unknown variance**: _Super_ common in AB testing. This is how boolean metrics are evaluated in ~every platform (that uses frequentist). This _could_ be done by `binomtest` but everyone I've worked with on this topic looks for the "z test".
2. **1-sample Z-test, with known variance**: Most straightforward application of the Z-test, I don't think this is under question.
3. **1-sample Z-Test of Proportions, with known variance**: Also a straightforward application of Z-test. Also not under question, I think?
4. **1-sample and 2-sample Z-test, with unknown variance**: Users could just use a T-Test instead and get the same results (for large samples, which is almost always the case in industry) , so the value of adding this flavor is not as high.
5. **2-sample Z-test, with known variance**: Not sure if this exists, need to read up on it.
6. **2-sample Z-Test of Proportions, with known variance**: Not sure if this exists either, need to read up on it.

I think these are all worth adding (except maybe 5 and 6), given the status Scipy has in the community and how common these tests are. I can't speak for the tradeoff with effort, I do agree that usefulness should outweigh completeness. I'm happy to do my part! Let me know how you'd like me to proceed!

**Comment by mdhaber** [comment]
@angadgill this is super useful. I have some questions.

> 1. **2-sample Z-Test of Proportions, with unknown variance**: _Super_ common in AB testing. This is how boolean metrics are evaluated in ~every platform (that uses frequentist). This _could_ be done by `binomtest` but everyone I've worked with on this topic looks for the "z test".

This is what you were suggesting with `ztest_ind_prop`, right? The arguments could be `k1`, `n1`, `k2`, `n2`, right? Two questions about that: 
- How would this be done by `binomtest`? That is a one-sample test of proportions.
- "two floats, one each for the two sides." How could `ztest_ind_prop` just accept two floats? The total number of observations will affect the p-value, right?

Can you describe a real-world example problem for this and how the data is collected? This sounds similar to Barnard's test or Boschloo's test. Why aren't those sufficient?

> 2. **1-sample Z-test, with known variance**: Most straightforward application of the Z-test, I don't think this is under question.

Great - can you describe an example problem and how you know the variance but not the mean?

> 3. **1-sample Z-Test of Proportions, with known variance**: Also a straightforward application of Z-test. Also not under question, I think?

This is exactly what `binomtest` tests, but it would be approximate instead of exact, right? It would just be approximating the binomial distribution with the normal distribution, right - a Wald test? Why would that be used instead of the binomial test, possibly with a "mid-p" correction?

> 4. **1-sample and 2-sample Z-test, with unknown variance**: Users could just use a T-Test instead and get the same results (for large samples, which is almost always the case in industry) , so the value of adding this flavor is not as high.

Well, yes, you would get almost exactly the same results with large samples, but the t-test will give you more accurate results for any sample size.

> 5. **2-sample Z-test, with known variance**: Not sure if this exists, need to read up on it.

It does, in the sense that I described the statistic and null distribution in https://github.com/scipy/scipy/issues/13662#issuecomment-3501071447 (even though I couldn't find it in one of thereferences). But that is almost certainly what the z.test is doing when you give it `x`, `y`, `sigma.x`, and `sigma.y`.

**Comment by angadgill** [comment]
Thank you for all your thoughtful questions, @mdhaber! Please see my answers below. This is a very long post, here's a TLDR:

- All the exact tests you've mentioned seem like good alternatives. I wasn't familiar with them so I'm not sure why Z-Test is often preferred over these. Statistical power maybe? I'll do some searching for an answer.
- Provided an example for Z-Test of Proportions.
- Couple of corrections from my last comment that you've caught, thanks!

---

>> 1. 2-sample Z-Test of Proportions, with unknown variance: Super common in AB testing. This is how boolean metrics are evaluated in ~every platform (that uses frequentist). 

> This is what you were suggesting with ztest_ind_prop, right? The arguments could be k1, n1, k2, n2, right?

That's right! Yes, `k1`, `n1`, `k2`, `n2` will work, better than my "two floats" suggestion actually, see below.

> How would this be done by binomtest? That is a one-sample test of proportions.

Good catch. I meant to add this next to "1-sample Z-Test of Proportions, with known variance" (late night comment sorry).

For the 2-sample Z-Test of Proportions, with unknown variance, folks in the industry would either:

- Use `statsmodels.stats.proportion.proportions_ztest` or 
- Use a chi2 test, example: [this website](https://www.evanmiller.org/ab-testing/chi-squared.html) (popular amongst AB testing practitioners) or
- Implement it themselves using `scipy.stats.norm` (or have ChatGPT implement it for them).

> "two floats, one each for the two sides." How could ztest_ind_prop just accept two floats? The total number of observations will affect the p-value, right?

Oops! You're right, we do need the total number of observations. `k1`, `n1`, `k2`, `n2` will work better here.

> Can you describe a real-world example problem for this and how the data is collected?

Classic use case in the industry for this is when we have _proportional metrics_ in an AB test.

**Example for 2-Sample Z-Test of Proportions**
We have a hypothesis that the replacing the existing blue button on our web app with a red button will increase the number of users who click on it. To test this, we randomize users who visit our web app into two groups: _control_ and _treatment_. We show the (existing) blue button to all users in the _control_ group and the red button to all users in the _treatment_ group. Our _metric_ is `% of users who clicked on the button`. Since each user is independent (due to our randomization) we can treat each user as a _trial_ (`n`) and clicking on the button as a _success_ (`k`) (multiple clicks from the same user count as 1 success).

In this example, the population variance is not known. Perhaps one could argue that we know the population variance by looking at historical data fro before the AB test? I'm not sure. 

>  This sounds similar to Barnard's test or Boschloo's test. Why aren't those sufficient?

I'm not familiar with these tests. Based on my cursory read, I think these will work too! Some thoughts (based on my _very_ quick read on Wiki):

- It seems like the exact tests (Barnards's, Boschloo's, and Fisher's) are more common for small samples.
- Z-Test of Proportions might be more powerful, based on [Wiki](https://en.wikipedia.org/wiki/Boschloo%27s_test) (see section titled _Comparison to other exact tests_). I don't access to the paper this is referencing, unfortunately.

---

>> 2. 1-sample Z-test, with known variance: Most straightforward application of the Z-test, I don't think this is under question.

> Great - can you describe an example problem and how you know the variance but not the mean?

Both, the population variance and mean, would be known in this case. Example: see page 5 in [Lecture 5: Hypothesis testing: One- and two-sample cases by Ernesto F. L. Amaral, Texas A&M University](https://www.ernestoamaral.com/docs/soci600-22fall/Lecture05.pdf?utm_source=chatgpt.com). Perhaps I should've written this as "1-Sample Z-Test, with known variance and mean".

---

>> 3. 1-sample Z-Test of Proportions, with known variance: Also a straightforward application of Z-test. Also not under question, I think?

> This is exactly what binomtest tests, but it would be approximate instead of exact, right? It would just be approximating the binomial distribution with the normal distribution, right - a Wald test? Why would that be used instead of the binomial test, possibly with a "mid-p" correction?

Yes I think the `binomtest` would work and yes Z-Test would be an approximation (but give the same results for large sample sizes). TBH, I'm not sure why Z-Test of Proportions is more popular (in my circles). Maybe it has more power? Your "mid-p" correction point is interesting and related to power. I'll do some searching and see if I can find some answers.

**Comment by mdhaber** [comment]
Regarding the two-sample test of proportions, what is a realistic click rate, sample size, and difference in click rate that you would consider meaningful? I can try to put together a comparison. 

One observation to throw in here - the two-sample `proportions_ztest` for equal proportions is equivalent to `chi2_contingency` with `correction=False`. So SciPy essentially has this capability already. 
<details>

```python3
import numpy as np
from scipy import stats, special
from statsmodels.stats.proportion import proportions_ztest

k1, k2 = 3, 4
n1, n2 = 4, 7

# with statsmodels 
z1, pval1 = proportions_ztest([k1, k2], [n1, n2])

# with SciPy
table = np.asarray([[k1, n1-k1], [k2, n2 - k2]]).T
res = stats.chi2_contingency(table, correction=False)
chi2, pval2 = res.statistic, res.pvalue

# manual
p1 = k1/n1
p2 = k2/n2
p = (k1 + k2) / (n1 + n2)
z3 = (p1 - p2) / np.sqrt(p * (1 - p) * (1/n1 + 1/n2))
pval3 = 2*special.ndtr(-z3)

print(z1, z3)  # 0.5922544268491786 0.5922544268491786
print(pval1, pval2, pval3)  # 0.553680228249833 0.553680228249833 0.553680228249833
```

</details>

---

> In this example, the population variance is not known. Perhaps one could argue that we know the population variance by looking at historical data fro before the AB test? I'm not sure.

For binomial data, the variance is a function of the number of trials and the probability of "success". So technically, if you know the number of trials, saying that you "know" the population variance is equivalent to saying that you know the probability of success. That said, we get the variance of the normal distribution in these tests by plugging in a pooled estimate of the probability of success as though it were the true, unknown value. So, I suppose in the denominator of the manual calculation above you could take `p` as the probability of success averaged over all the past data, too. 

---

> Both, the population variance and mean, would be known in this case. Example: see page 5

I was hoping for a real example from your experience in which the variance is considered to be known. On page 5, it is assumed that the sample from the community is large enough for the standard deviation to be considered "known". That might not change the conclusions of a test, but it wouldn't be so much harder to do the "correct" thing - just treat it as a two-sample t-test with one group much larger than the other.

<details>

```python3
from scipy import stats, special

mean1 = 6.8  # sample mean
n1 = 127  # sample size

mean2 = 7.2  # "known" mean from community
n2 = np.asarray([127, 1270, 12700, 127000, 1270000])  # community sample size

std = 1.43  # "known" standard deviation

res = stats.ttest_ind_from_stats(mean1, std, n1, mean2, std, n2)
print(res.pvalue)  # [0.02669582 0.00269801 0.0017128  0.0016291  0.00162088]

z = (mean1 - mean2) / (std / n1**0.5)
print(special.ndtr(z)*2)  # 0.001619967073266778
```
Note that unless the community sample size is really small, the z-test approximation is not very accurate.

</details>

**Comment by angadgill** [comment]
> Regarding the two-sample test of proportions, what is a realistic click rate, sample size, and difference in click rate that you would consider meaningful? I can try to put together a comparison.

Couple of examples from experience on two very different products and metrics, but mostly made-up here:

- Proportion metric has a baseline of 80% (historic average) and we want to measure an uplift of 1% (relative). This would need [~40k samples](https://www.evanmiller.org/ab-testing/sample-size.html#!80;80;5;1;1) per variant. 
- Proportion metric has a baseline of 5% and we want to measure an uplift of 10% (relative). This would need [~30k samples](https://www.evanmiller.org/ab-testing/sample-size.html#!5;80;5;10;1) per variant.

> One observation to throw in here - the two-sample proportions_ztest for equal proportions is equivalent to chi2_contingency with correction=False. So SciPy essentially has this capability already.

Oh wow, those are outputs identical. That's fantastic! I am also a bit surprised  because I would expect  `ztest` to be anti-conservative with low sample size. Though maybe that's true for `chi2_contingency` with `correction = False`? This is enlightening, thank you!

I wonder if we should write a doc that redirects wayward statistics enthusiasts towards Chi2, Barnards, Boschloo, etc.? Or perhaps write a Z-Test of Proportions that's just a wrapper around `chi2_contingency`? (or both?)

---

> I was hoping for a real example from your experience in which the variance is considered to be known. 

Sorry, I've actually never seen this come up in AB Testing. Perhaps the sample sizes are large enough that it doesn't make a meaningful difference so it doesn't get talked about, not sure.

> On page 5, it is assumed that the sample from the community is large enough for the standard deviation to be considered "known". 

I'm curious about this distinction. I thought that's what "known" meant!

**Comment by rkern** [comment]
> I wonder if we should write a doc that redirects wayward statistics enthusiasts towards Chi2, Barnards, Boschloo, etc.? Or perhaps write a Z-Test of Proportions that's just a wrapper around `chi2_contingency`? (or both?)

Documentation is, I think, the preferred way to handle this case. We always want to reduce the number of aliases. We have lots of notes in our documentation, though, like "this is also sometimes known as Method X".

> I'm curious about this distinction. I thought that's what "known" meant!

No, "known" means, well, "known", or at least "assumed", but in any case "not estimated from the samples". When the sample sizes get large, the computations of abusing the Z-test with estimated variances converge with those of more rigorous tests designed for estimated variances, so one can often get away with it. However, "can often get away with it" is not the overall direction we generally want to take in designing our `scipy.stats` APIs. This is why we're going over these use cases with a fine-toothed comb. If we have preferred options that are more robust, we probably don't want to add Z-test implementations that cover those same cases.

It's my impression, from experience in a variety of domains, that Z-test is largely a pedagogical tool. It's simple enough for students to get the hang of all the rest of the confusing semantics of hypothesis testing without extra complicating details. But in practice, its assumptions are almost always violated. But it persists in the minds of students who go on to be experts in domains aren't statistics, and reach for it when designing their protocols and don't get caught out because they just happen to have sample sizes that are enough that it doesn't really make much of a difference to the results. There are almost always better alternatives that improve the robustness of the overall protocols to low-sample-size scenarios automatically, with little cost to runtime.

Overall, our goal for `scipy.stats` is not completeness or pedagogy (both are fine goals, though, for other projects). If anything, it's "guiding people to use good defaults". Or more practically, "what gets the most community benefit out of our, especially @mdhaber's, bounded time to maintain all this".

**Comment by josef-pkt** [comment]
law of large number and central limit theorem work whether variance or other parameters are estimated or assumed known.

Those just rely on the usual "technical" assumption, existence of moments (e.g. up to 4th), domain independent of parameters, differentiability, ....
Even if some of those assumptions don't hold (e.g. differentiable in median), we still get law of large numbers and asymptotic normal distribution but not of the simple standard form or not with standard proof.

So, if "we can get away with it" only depends on that the sample size is large enough.

All models outside of the linear models, Poisson, GLM, Beta regression, failure models,... use asymptotic normal distribution for inference.
In a few cases "exact" methods are also available.

For me "guiding people to use good defaults" means offering alternatives for assumptions that have a large impact like misspecified variance, e.g Welch t-test and Welch anova instead of the ones that assume equal variances.
Those have large effects, while normal versus t (or chisquare versus F) usually does not have much impact in "reasonable" sample sizes. And often the theory has to rely on asymptotic normal distribution. 

For example 2-sample proportion test with better small sample properties:
Those make different adjustments to the test statistic to improve small sample behavior, but they all rely on the normal distribution for the p-value. (Also because we assume that we know the relationship between variance and mean.) (*)
https://www.statsmodels.org/dev/_modules/statsmodels/stats/proportion.html#test_proportions_2indep


But in any case, z-tests are in already in statsmodels.


(*) AFAIK, there is no theoretical justification for using the t-test in binomial proportions or Poisson rates, even if it "works".
There are other asymptotic derivations like large-deviation asymptotics or higher order asymptotic expansions, but I'm not very familiar with them and never implemented any of those methods.

**Comment by angadgill** [comment]
Hey @mdhaber! Just following up here for closure. I believe we've decided to not add new code for Z-Test and instead update the docs. I'm happy to make that update! Any preference on where this should go? I'm guessing it would go under "User Guide", perhaps on [Analysing one sample](https://docs.scipy.org/doc//scipy/tutorial/stats/analysing_one_sample.html#t-test-and-ks-test) and [Comparing two samples](https://docs.scipy.org/doc//scipy/tutorial/stats/comparing_two_samples.html)? 

[Bit of a tangent: Scipy Docs sometimes are verrryy slows to load. I don't see any issues about this. Maybe its on my end]

**Comment by mdhaber** [comment]
Re: docs, see gh-23757.
Let's start with docs, yeah. Those tutorials would be ok places if what you want to write fits in nicely, but no need to be constrained by them. I think it would be OK to write a separate tutorial, the common theme being what to do if you think you need a z-test.

**Comment by angadgill** [comment]
Hey @mdhaber! Finally coming around to this. 

I'm think it would be good to create a new paged called "Z-Test"  under [Hypothesis tests](https://docs.scipy.org/doc//scipy/tutorial/stats/hypothesis_tests.html). Here's the outline I'm thinking for it:
- Top paragraph describing: what the z-test is and that it's not available in scipy
- A table showing Z-test of means alternatives: 

One-sample / Two-sample | Known / unknown variance | Alternative offered
--- | --- | ---
One-sample | Known variance | Use `norm` directly
One-sample | Unknown variance | `ttest_1samp`
Two-sample | Known variance | Use `norm` directly
Two-sample | Unknown variance | `ttest_ind`, `test_ind_from_stats`

- A table showing Z-test of proportions alternatives:

One-sample / Two-sample |  Alternative offered
--- | --- 
One-sample | `binomtest`
Two-sample | `chi2_contingency (correction=False)`, `barnard_exact`, `boschloo_exact`

- Examples (with code) for all use cases shown above.
 

If this looks okay to you, I'll go ahead and write it + put up a PR. Thank you!

**Comment by mdhaber** [comment]
That looks OK, but I would probably not be the one to review it anymore. @fbourgey, you've been working on some stats tutorial improvements. Would you be willing to review this?
